app/bundles/EmailBundle/EventListener/MatchFilterForLeadTrait.php

```diff
@@ -0,0 +1,187 @@
+<?php
+/*
+ * @copyright   2017 Mautic Contributors. All rights reserved
+ * @author      Mautic
+ *
+ * @link        http://mautic.org
+ *
+ * @license     GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
+ */
+
+namespace Mautic\EmailBundle\EventListener;
+
+/**
+ * Trait MatchFilterForLeadTrait.
+ */
+trait MatchFilterForLeadTrait
+{
+    /**
+     * @param array $filter
+     * @param array $lead
+     *
+     * @return bool
+     */
+    protected function matchFilterForLead(array $filter, array $lead)
```

Comment:
Very long and complicated method. Do we have a tests for use cases? It is not part of this PR.